README
=======
README
=======
Firstly let me thank you for deciding to try netmon, if you
have any problems please look at the FAQ if that fails you can 
contact me by email or at the projects homepage.

Project homepage - https://sourceforge.net/projects/netmonitor/
Contact Author: hellcat82@iprimus.com.au

ABOUT
------
netmon is a network monitoring console application which shows
the bandwidth consumed and the total transmitted and recieved
data. It also logs this data so that it can retain these values
after a crash or restart. This makes netmon handy for people who
have broadband connections that require payment per download
volume.

Its main features are
-Shows transfer rates and data totals with dynamic units
-Ability to force units
-Logs the data totals so that it can retain these values
 after restarts or crashes
-Low resource usage
-Output fits on a 66x9 terminal screen
-Ability to log up to 20971520TB

COMPILING
------------
To compile netmon type the following command
./configure
./make
./make install

For more information please see the INSTALL file

INSTALLING
------------
Please see INSTALL file

UNINSTALLING
-------------
To uninstall netmon you need root priviliges, unistall with
make uninstall

MANUAL DATA SETUP
-----------------
Due to the change in log file format this method is now 
depreciated. Please run netmonitor with the --config flag

RUNNING NETMON
-------------
To run netmon type the following
netmonitor

To find out more about netmon type
netmonitor --help or man netmonitor


NEWS
====
NEWS
====
For the latest news please visit http://netmonitor.sourceforge.net


src/netmon.c
/*****************************************************************************
 *netmon.c:  Main program for netmonitor, netmonitor is a network monitoring
 *           and logging program for linux. The aim is to provide a bandwidth
 *           meter type program that is small and simple
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 05 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 ****************************************************************************/


#include "net_lib.h"


int main(int argc, char *argv[]){

  /*File Paths*/
  char filepath[] = "/proc/net/dev";

  char *homepath = getenv("HOME");

  char logpath[80];
  strcpy(logpath, homepath);
  strcat(logpath, "/.netmonitor/data.dat");

  char confpath[80];
  strcpy(confpath, homepath);
  strcat(confpath, "/.netmonitor/device.conf");


  /*File pointers*/
  FILE *dev_file = NULL;
  FILE *log_file = NULL;
  FILE *conf_file = NULL;

  net_t *data;
  int args;                       /*Arg return value*/
  char t_buff[160];               /*Temporary buffer*/
  char mod_dev[160];              /*Buffer for dev modifications*/
  int size = 0;                   /*Size of the net_t array*/
  int index = 0;                  /*Index to device array entry*/
  char* delim;                    /*Pointer to end of device*/
  long long temprx, temptx;       /*Stores dev values before update*/
  time_t start, finish;           /*Used for timing data*/
  int first_run = 1;              /*Check for program start*/

  time(&start);
  if((args = parse_args(argc, argv)) == 1){
    config(filepath, homepath);
    exit(EXIT_SUCCESS);
  }
  if(args == -1)
    exit(EXIT_SUCCESS);

  /***************************CONFIG FILE ACCESS***************************/

  if((conf_file = fopen(confpath, "r")) == NULL){
    fprintf(stderr, "Error: Unable to open config file. ");
    fprintf(stderr, "Please run with --config flag\n");
    exit(EXIT_FAILURE);
  }

 if(! fgets(t_buff, sizeof(t_buff), conf_file)){
    fprintf(stderr, "Error: Unable to read config file\n");
    exit(EXIT_FAILURE);
  }
int j, i = atoi(t_buff);
  if(DEBUG){
    fprintf(stderr, "main:CONFIG FILE ACCESS: number of devices in ");
    fprintf(stderr, "config file%d\n", i);
  }
  data = calloc(i, sizeof(net_t));
  for(j = 0; j < i; j++){
    if (!fgets(t_buff, sizeof(t_buff), conf_file)){
      fprintf(stderr, "Error: Recieved end of file before expected\n");
      exit(EXIT_FAILURE);
    }
    remove_newline(t_buff);
    strcpy(data[j].device, t_buff);
    if(DEBUG){
      fprintf(stderr, "main:CONFIG FILE ACCESS: device entered in");
      fprintf(stderr, " data[%i] is %s\n", j ,data[j].device);
    }
  }
  size = i;
  if(DEBUG)
    fprintf(stderr, "main:CONFIG FILE ACCESS: Size of array is %d\n", size);
  fclose(conf_file);

  /*****************************LOG FILE ACCESS V0.5***********************/
  long long rxbval[5];
  long long txbval[5];
  long long rxgval[5];
  long long txgval[5];
  long long clearrxval[5];
  long long cleartxval[5];

  /*Open log file for reading*/
  if((log_file = fopen(logpath, "r")) == NULL){
    fprintf(stderr, "Error: Unable to open log file. ");
    fprintf(stderr, "Please run with --config flag\n");
    exit(EXIT_FAILURE);
  }

  /*Checks log file version*/
  if(!fgets(t_buff, sizeof(t_buff), log_file)){
    fprintf(stderr, "Error: Unable to read log file\n");
  }
  if(strstr(t_buff, "<0.5>") == NULL){
    fprintf(stderr, "Error: Wrong log file version. Please update\n");
    exit(EXIT_FAILURE);
  }
  /*******LOG file loop*********/
  for(i = 0; i < size; i++){
    if(DEBUG)
      fprintf(stderr, "main:LOG FILE ACCESS: Attempting to read log file\n");

    if(!fgets(t_buff, sizeof(t_buff), log_file)){
      fprintf(stderr, "Error: Unable to read log file\n");
    }

    /*Find correct device*/
    for(index = 0; index < size; index ++){
      if(strstr(t_buff, data[index].device) != NULL){
	break;
      }
    }
    if(DEBUG)
      fprintf(stderr, "main:LOG FILE ACCESS: Found device at position %d\n",
	      index);
    /*Reads in rx byte value from log*/
    if(!fgets(t_buff, sizeof(t_buff), log_file)){
      fprintf(stderr, "Error: Unable to retrieve data values\n");
      exit(EXIT_FAILURE);
    }
    sscanf(t_buff, "%lld", &rxbval[index]);

    /*Reads in rx gig value from log*/
    if(!fgets(t_buff, sizeof(t_buff), log_file)){
      fprintf(stderr, "Error: Unable to retrieve data values\n");
      exit(EXIT_FAILURE);
    }

    sscanf(t_buff, "%lld", &rxgval[index]);



    /*Reads in tx byte value from log*/
    if(!fgets(t_buff, sizeof(t_buff), log_file)){
      fprintf(stderr, "Error: Unable to retrieve data values\n");
      exit(EXIT_FAILURE);
    }
    sscanf(t_buff, "%lld", &txbval[index]);

    /*Reads in tx gig value from log*/
    if(!fgets(t_buff, sizeof(t_buff), log_file)){
      fprintf(stderr, "Error: Unable to retrieve data values\n");
      exit(EXIT_FAILURE);
    }
    sscanf(t_buff, "%lld", &txgval[index]);
  }

  fclose(log_file);

    /***************************DEV FILE ACCESS****************************/
  /*Check dev file*/
  if((checkfile(filepath)) != 0){
    fprintf(stderr, "File security failed\n");
    exit(EXIT_FAILURE);
  }
  if(DEBUG)
    fprintf(stderr, "main:DEV FILE ACCESS:dev file passed security check\n");

  /*Open dev file for reading*/
  if((dev_file = fopen(filepath, "r")) == NULL){
    fprintf(stderr, "Error: Unable to open dev file\n");
    exit(EXIT_FAILURE);
  }

  /*Check dev file version*/
  if((ver_check(dev_file)) < 0){
    fprintf(stderr, "Error: Incorrect dev version\n");
    exit(EXIT_FAILURE);
  }
  if(DEBUG){
    fprintf(stderr, "main:DEV FILE ACCESS: Dev file is the correct ");
    fprintf(stderr, "version\n");
  }
  rewind(dev_file);
  if(DEBUG)
    fprintf(stderr, "main:DEV FILE ACCESS: Dev file is at the beginning\n");

  /*******************************UPDATE LOOP******************************/
  while(1){
    if(DEBUG)
      fprintf(stderr, "main:UPDATE LOOP: Entering update loop now\n");

    /*removes headings*/
    for(i = 0; i < 1; i++){
      if( !fgets(t_buff, sizeof(t_buff), dev_file)){
	fprintf(stderr, "Error:UPDATE LOOP Unable to read dev file\n");
	exit(EXIT_FAILURE);
      }
    }
    int found = 0;
    /*Checks that device is in data*/
    while(fgets(t_buff, sizeof(t_buff), dev_file)){
      found = 0;
      for(j = 0; j < size; j++){
	if(strstr(t_buff, data[j].device) != NULL){
	  found = 1;
	  if(DEBUG){
	    fprintf(stderr, "main:UPDATE LOOP:Found matching dev file ");
	    fprintf(stderr,"device at position %d\n",j);
	  }
	  break;
	}
      }
      if(DEBUG){
	fprintf(stderr, "main:UPDATE LOOP: Exited out of update loops ");
	fprintf(stderr, "device check\n");
      }
      if(found){
	if(DEBUG)
	  fprintf(stderr, "main:UPDATE LOOP: Entered data update section\n");
        /*Retrieves values*/
	delim = strchr(t_buff, ':');

	if( *( delim + 1 ) == ' ' ){
	  sscanf(t_buff,"%*s %Lu %*d %*d %*d %*d %*d %*d %*d %Lu ",
		 &temprx, &temptx);
	  if(first_run){
	      clearrxval[j] = temprx;
	      cleartxval[j] = temptx;
	    }
	}
	else{
	  if(DEBUG){
	    fprintf(stderr, "main:UPDATE LOOP: Found no space between ");
	    fprintf(stderr,"device and value\n");
	  }
	  strcpy(mod_dev ,delim + 1);
	  for(i = 0; i < sizeof(mod_dev); i++){
	    if(mod_dev[i] == ' '){
	      mod_dev[i] = '\0';
	      break;
	    }
	  }
	  temprx = atoll(mod_dev);
	  sscanf(t_buff,"%*s %*d %*d %*d %*d %*d %*d %*d %Lu ", &temptx);
	  if(first_run)
	    {
	      clearrxval[j] = temprx;
	      cleartxval[j] = temptx;
	    }
	}
	if(first_run){
	  set_tx(&data[j], cleartxval[j], txbval[j], txgval[j]);
	  set_rx(&data[j], clearrxval[j], rxbval[j], rxgval[j]);
	}
	else{
	  update_rx(&data[j], temprx);
	  update_tx(&data[j], temptx);
	}
      }
    }
    first_run = 0;
    rewind(dev_file);
    if(DEBUG)
      fprintf(stderr, "main:UPDATE LOOP: dev file rewinded\n");
    time(&finish);
    if(DEBUG)
      fprintf(stderr, "main:UPDATE LOOP: start time is %d seconds\n", (int)start);
    if(DEBUG)
      fprintf(stderr, "main:UPDATE LOOP: finish time is %d seconds\n", (int)finish);
    time(&finish);
    display(data, size,(double)(finish - start));
    time(&start);
    if((write_to_log(data, logpath, size)) < 0){
      fprintf(stderr, "main:UPDATE LOOP: Error writing to log. Exiting!");
      exit(EXIT_FAILURE);
    }
    sleep(2);
  }
  exit(EXIT_SUCCESS);
}



src/net_lib.h
/*******************************************************************
 *net_lib.h: Function definitions to be used with netmon.c
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 09 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 *******************************************************************/
#ifndef _NET_LIB
#define _NET_LIB

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sys/file.h>
#include <time.h>
#include <pthread.h>

#include "net_data.h"
#include "config_lib.h"

#define DEBUG 0

/*******************************************************************
 *void display(): Prints values stored in Ndata to the screen with
 *                a table like format, before each print to screen
 *                it flushes the console.
 *
 *Recieves:       net_t array, int size of the array, and a double
 *                time value
 *Returns:        NA
 */
void display(net_t *Ndata, int size ,double time);

/*******************************************************************
 *void clearsrceen(): Clears previous text from the console screen
 *
 *Recieves:           NA
 *
 *Returns:            NA
 */
void clearscreen();
/*******************************************************************
 *int parse_args(): Parses the command line arguments and
 *                  performs the appropriate operations for the
 *                  arguments.
 *
 *Recieves:         int number of arguments, char pointer to
 *                  argument vector.
 *
 *Returns:          A int value which tells the caller what
 *                  argument was found
 */
int parse_args(int argc, char* argv[]);

/*******************************************************************
 *long double to_unit(): Modifies a long number to a larger unit
 *                       and changes the it's unit value to that
 *                       unit
 *
 *Recieves:              char* to a string holding the units
 *                       long long value to be modified
 *
 *Returns:               long double value that is the modified
 *                       long long value
 */
long double to_unit(char *unit, long double bval, long double gval ,
		    int mode);

/*******************************************************************
 *int checkfile(): Checks that a file isn't a symbolic link for
 *                 security reasons
 *
 *Recieves:        char* holding the path to the file to be checked
 *
 *Returns:         int value -1 if check failed, 0 if check passed
 */
int checkfile(char *path);

/*******************************************************************
 *int ver_check(): Checks that the dev file is the correct version.
 *
 *Recieves:        FILE* to the dev file stream.
 *
 *Returns:         int value -1 if version was incorrect, 0
 *                 otherwise.
 */
int ver_check(FILE *dev_file);

/*******************************************************************
 *void remove_newline(): Removes first new line from a string
 *
 *Recieves:              char* to the string to be modified
 *
 *Returns:               NA
 */
void remove_newline(char* string);

/*******************************************************************
 *int write_to_log(): Writes data from net_t array to a file
 *
 *Recieves:           net_t array holding the data to be written
 *                    char* to the path of the file
 *                    int size of the array
 *
 *Returns:            int value -1 if write failed, 0 otherwise
 */
int write_to_log(net_t *Ndata, char* path, int size);

/******************************************************************
 *void display_help(): Displays the help screen.
 *
 *Recieves:            NA
 *
 *Returns:             NA
 */
void display_help();

#endif
#include <time.h>


src/net_data.c
/************************************************************
 *net_data.c: Functions for the functions and struct in
 *            net_data.h, to be used with netmon.c
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 09 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 ************************************************************/
#include "net_data.h"

void init_net_t(net_t *Ndata){
  Ndata->gig_rx = 0;
  Ndata->pre_rx = 0;
  Ndata->byte_rx = 0;
  Ndata->lval_rx = 0;
  Ndata->gig_tx = 0;
  Ndata->pre_tx = 0;
  Ndata->byte_tx = 0;
  Ndata->lval_tx = 0;
}
/**********************SET FUNCTIONS************************/
void set_rx(net_t *Ndata, long long devval ,long long bval,
	    long long gval){
  Ndata->pre_rx = bval;
  Ndata->byte_rx = bval;
  Ndata->gig_rx = gval;
  Ndata->lval_rx = devval;
}

void set_tx(net_t *Ndata, long long devval ,long long bval,
	    long long gval){
  Ndata->pre_tx = bval;
  Ndata->byte_tx = bval;
  Ndata->gig_tx = gval;
  Ndata->lval_tx = devval;
}

/**********************GET FUNCTIONS**************************/
long long get_rxRate(net_t *Ndata, double elapse){
  if(elapse == 0){
    elapse = 1;
  }
  return (Ndata->byte_rx - Ndata->pre_rx)/elapse;
}

long long get_txRate(net_t *Ndata, double elapse){
  if(elapse == 0){
    elapse = 1;
  }
  return (Ndata->byte_tx - Ndata->pre_tx)/elapse;
}

long long get_rxbtotal(net_t *Ndata){
  return Ndata->byte_rx;
}

long long get_txbtotal(net_t *Ndata){
  return Ndata->byte_tx;
}

long long get_rxgtotal(net_t *Ndata){
  return Ndata->gig_rx;
}

long long get_txgtotal(net_t *Ndata){
  return Ndata->gig_tx;
}
/**********************UPDATE FUNCTIONS***********************/
void update_tx(net_t *Ndata, long long val){
  // Check if new value is less than old value
  long long tval;
  if(val < Ndata->lval_tx){
    tval = val + (long long)858993459 * (long long)5;
  }
  else{
    tval = val;
  }
  //Set tval
  tval = tval - Ndata->lval_tx;
  //reset lval
  Ndata->lval_tx = val;
  //Update values
  Ndata->pre_tx = Ndata->byte_tx;
  Ndata->byte_tx = Ndata->byte_tx + tval;
  //Check if pre is 1 gig yet
  //Normalizes 1 gig each cycle so no need to loop
  if(Ndata->pre_tx > ND_ONEGIG){
    Ndata->pre_tx = Ndata->pre_tx - ND_ONEGIG;
    Ndata->byte_tx = Ndata->byte_tx - ND_ONEGIG;
    Ndata->gig_tx++;
  }
}

void update_rx(net_t *Ndata, long long val){
  // Check if new value is less than old value
  long long tval;
  if(val < Ndata->lval_rx){
    tval = val + (long long)858993459 * (long long)5;
  }
  else{
    tval = val;
  }
  //Set tval
  tval = tval - Ndata->lval_rx;
  //reset lval
  Ndata->lval_rx = val;
  //Update values
  Ndata->pre_rx = Ndata->byte_rx;
  Ndata->byte_rx = Ndata->byte_rx + tval;
  //Check if pre is 1 gig yet
  //Normaizes 1 gig each cycle so no need to loop
  if(Ndata->pre_rx > ND_ONEGIG){
    Ndata->pre_rx = Ndata->pre_rx - ND_ONEGIG;
    Ndata->byte_rx = Ndata->byte_rx - ND_ONEGIG;
    Ndata->gig_rx++;
  }
}


src/net_data.h
/***********************************************************
 *net_data.h: Contains the net_t struct and it's functions
 *            for use with netmon.c
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 09 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 ***********************************************************/
#ifndef NET_DATA
#define NET_DATA
#define ND_ONEGIG 1073741824

#include <sys/types.h>
#include <string.h>
typedef struct net_data{
  char device[6];
  long long byte_rx; //Holds bytes rx value
  long long gig_rx;  //Holds the gigabyte rx value
  long long pre_rx;  //Holds the last rx bytes count
  long long lval_rx; //Holds the last dev rx value
  long long byte_tx; //Holds bytes tx value
  long long gig_tx;  //Holds the gigabyte tx count
  long long pre_tx;  //Holds the last tx bytes count
  long long lval_tx; //Last dev tx value
} net_t;

/************************************************************
 *void init_net_t(): Initiates all values of a net_t
 *                   structure
 *
 *Recieves:          net_t* to the structure to be set
 *
 *Returns:           NA
 */
void init_net_t(net_t *Ndata);

/************************************************************
 *void set_tx:  Sets the rx value of the net_t structure
 *
 *Recieves:     *Ndata structure
 *              *devval: initial value from the dev file
 *              *bval: byte value from log file
 *              *gval: gig value from log file
 *
 *Returns:      NA
 */
void set_rx(net_t *Ndata, long long devval,  long long bval,
	    long long gval);

/************************************************************
 *void set_tx:  Sets the tx value of the net_t structure
 *
 *Recieves:     *Ndata structure
 *              *devval: initial value from the dev file
 *              *bval: byte value from log file
 *              *gval: gig value from log file
 *
 *Returns:      NA
 */
void set_tx(net_t *Ndata, long long devval, long long bval,
	    long long gval);
/***********************************************************
 *long long get_rxbtotal(): Returns the total rx bytes
 *
 *Recieves:   Ndata structure
 *
 *Returns:   long long value of rx bytes
 */
long long get_rxbtotal(net_t *Ndata);

/***********************************************************
 *long long get_txbtotal(): Returns the total tx bytes
 *
 *Recieves:   Ndata structure
 *
 *Returns:    long long value of tx bytes
 */
long long get_txbtotal(net_t *Ndata);

/***********************************************************
 *long long get_rxgtotal(): Returns the total rx gigs
 *
 *Recieves:   Ndata structure
 *
 *Returns:    long long value of rx gigs
 */
long long get_rxgtotal(net_t *Ndata);

/***********************************************************
 *long long get_txgtotal(): Returns the total tx gigs
 *
 *Recieves:   Ndata structure
 *
 *Returns:    long long value of tx gigs
 */
long long get_txgtotal(net_t *Ndata);

/***********************************************************
 *long long get_rxRate(): Gets the rate of change in two
 *                        updated set_rx() calls per second
 *
 *Recieves:               net_t structure to get value from.
 *                        double elapsed time between the
 *                        two set_rx() calls
 *
 *Returns:                The rate of change per second
 */
long long get_rxRate(net_t *Ndata, double elapse);

/***********************************************************
 *long long get_txRate(): Gets the rate of change in two
 *                        updated set_tx() calls per second
 *
 *Recieves:               net_t structure to get value from.
 *                        double elapsed time between the
 *                        two set_tx() calls
 *
 *Returns:                The rate of change per second.
 */
long long get_txRate(net_t *Ndata, double elapse);

/**********************************************************
 *void update_rx(): Updates the total data counters for rx,
 *                  contains logic for detecting dev file
 *                  overflow
 *
 *Recieves:         * Ndata structure
 *                  * val: The data value from dev file
 *
 *Returns:          NA
 */
void update_rx(net_t *Ndata, long long val);

/**********************************************************
 *void update_rx(): Updates the total data counters for rx,
 *                  contains logic for detecting dev file
 *                  overflow
 *
 *Recieves:         * Ndata structure
 *                  * val: The data value from dev file
 *
 *Returns:          NA
 */
void update_tx(net_t *Ndata, long long val);
#endif


src/config_lib.c
/******************************************************************
 *config_lib.c : Provides function definitions for the header file
 *               config_lib.c. It is the functions for the
 *               configuration subprogram for netmonitor.
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 20 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 ******************************************************************/
#include "config_lib.h"

static int is_dev = 0;

void config(char* dev_path, char* home_path){
  FILE *dev;
  char t_buff[150];
  conf_t devs[7];
  int num_dev;        /*Total number of devices*/
  char c_buff;
  int i, j, iput, is_dev = 0;

  if((dev = fopen(dev_path, "r")) == NULL){
    fprintf(stderr, "Error: Unable to open device file for reading\n");
    exit(EXIT_FAILURE);
  }
  if(vercheck(dev) != 0){
    fprintf(stderr, "Error: Device file is wrong version\n");
    exit(EXIT_FAILURE);
  }
  /*Remove header*/
  for(i = 0; i < 1; i++){
    if( !fgets(t_buff, sizeof(t_buff), dev)){
      fprintf(stderr, "Error: Unable to read device file\n");
      exit(EXIT_FAILURE);
    }
  }
  /*Read in values*/
  i = j = num_dev = 0;
  while(fgets(t_buff, sizeof(t_buff), dev)){
    j = i = 0;
    while(t_buff[i] != ':'){
      if(t_buff[i] != ' '){
	devs[num_dev].dev[j] = t_buff[i];
	j++;
      }
      i++;
    }
    devs[num_dev].added = 0;
    devs[num_dev].dev[j] = '\0';
    num_dev ++;
  }

  /*Main interface loop*/
  while(1){
    clrscrn();
    display_main_config();
    c_buff = getc(stdin);

    while(getc(stdin) != '\n');
    if(isdigit(c_buff)){
      iput = atoi(&c_buff);

      switch(iput){
      case 1:
	conf_add(devs, num_dev);
	break;
      case 2:
	conf_remove(devs, num_dev);
	break;
      case 3:
	i = conf_save(devs, num_dev, home_path);
	if(i == 1){
	  exit(EXIT_SUCCESS);
	}
	else if(i == -1){
	  exit(EXIT_FAILURE);
	}
	break;
      case 4:
	fprintf(stdout, "Are you sure you want to exit without saving [y|N] : ");

	c_buff = getc(stdin);
	while(getc(stdin) != '\n');

	if(c_buff == 'y')
	  exit(EXIT_SUCCESS);
	else
	  break;
      }

    }
  }
}

void display_main_config(){
  fprintf(stdout, "Netmonitor main configuration menu\n\n");
  fprintf(stdout, "1 - Add device\n");
  fprintf(stdout, "2 - Remove device\n");
  fprintf(stdout, "3 - Save configuration file and exit\n");
  fprintf(stdout, "4 - Exit without saving\n");
  fprintf(stdout, "Enter choice: ");

}

void conf_add(conf_t *dev, int size){
  char c_buff[3];
  int i, val;
  while(c_buff[0] != 'e'){
    clrscrn();
    fprintf(stdout,"Add device\n");
    fprintf(stdout, "\nAvailable devices\n");
    for(i = 0; i < size; i++){
      if(dev[i].added == 0)
	fprintf(stdout, "%d - %s\n", i , dev[i].dev);
    }
    fprintf(stdout, "\nSelected devices\n");
    for(i = 0; i < size; i++){
      if(dev[i].added == 1)
	fprintf(stdout, "%s\n", dev[i].dev);
    }
    fprintf(stderr, "\nPlease select a device to add (type 'e' when finished): ");
    fprintf(stdout, " ");
    read(STDIN_FILENO, c_buff, sizeof(c_buff));

    if(isdigit(c_buff[0])){
      val = atoi(c_buff);
      if(val < size){
	if(dev[val].added == 0){
	  dev[val].added = 1;
	  is_dev ++;
	}
      }
    }
  }
}

void conf_remove(conf_t *dev, int size){
  char c_buff[3];
  int i, val;
  while(c_buff[0] != 'e'){
    clrscrn();
    fprintf(stdout,"Remove device\n");
    fprintf(stdout, "\nAvailable devices\n");
    for(i = 0; i < size; i++){
      if(dev[i].added == 0)
	fprintf(stdout, "%s\n", dev[i].dev);
    }
    fprintf(stdout, "\nSelected devices\n");
    for(i = 0; i < size; i++){
      if(dev[i].added == 1)
	fprintf(stdout, "%d - %s\n", i, dev[i].dev);
    }
    fprintf(stderr, "\nPlease select a device to remove (type 'e' when finished): ");
    fprintf(stdout, " ");
    read(STDIN_FILENO, c_buff, sizeof(c_buff));

    if(isdigit(c_buff[0])){
      val = atoi(c_buff);
      if(val < size){
	if(dev[val].added != 0){
	  dev[val].added = 0;
	  is_dev --;
	}
      }
    }
  }
}


int conf_save(conf_t *dev, int size, char* homepath){
  FILE *dev_conf;
  FILE *data_dat;
  int i, dev_count = 0;
  char c_buff;
  char device_path[80], data_path[80], netmon_path[80];
  char version[] = "<0.5>";
  if(is_dev == 0){
    fprintf(stderr, "Error: Nothing to save\n");
    return 1;
  }
  fprintf(stdout, "This will erase all previous data. Are you sure ? [y|N]");
  c_buff = getc(stdin);
  while(getc(stdin) != '\n');

  if(c_buff == 'y'){
    /*Creates Paths*/
    strcpy(device_path, homepath);
    strcpy(data_path, homepath);
    strcpy(netmon_path, homepath);
    strcat(device_path, "/.netmonitor/device.conf");
    strcat(data_path, "/.netmonitor/data.dat");
    strcat(netmon_path, "/.netmonitor");
    /*checks and creates directory*/
    if(access(netmon_path, F_OK) == -1)
      mkdir(netmon_path, S_IRWXU);

    if((dev_conf = fopen(device_path, "w+")) == NULL){
      fprintf(stderr, "Error: Could not create device file\n");
      return -1;
    }

    if((data_dat = fopen(data_path, "w+")) == NULL){
      fprintf(stderr, "Error: Cound not create data file\n") ;
      return -1;
    }
    /*Prints version number*/
    fprintf(data_dat, "%s\n", version);
    /*Print to data file*/
    for(i = 0; i < size; i++){
      if(dev[i].added == 1){
	dev_count++;
	fprintf(data_dat,"[%s]\n0\n0\n0\n0\n", dev[i].dev);
      }
    }
    /*Print to device.conf*/
    fprintf(dev_conf, "%d\n", dev_count);
    for(i = 0; i < size; i++){
      if(dev[i].added == 1){
	fprintf(dev_conf,"%s\n", dev[i].dev);
      }
    }
  }
  else
    return 2;

  return 1;
}

int vercheck(FILE *dev_file){
  char buff[80];
  int i = 0;
  for(i = 0; i < 2; i++){
    if(!fgets(buff, sizeof(buff), dev_file)){
      return -1;
    }
  }
  if((strstr(buff ,"compressed")) == NULL){
    return -2;
  }
  return 0;
}

void clrscrn(){
  char clear[3];
  clear[0] = 033; clear[1] = 'c'; clear[2]='\0';
  fprintf(stdout, "%s", clear);
}


src/net_lib.c
/***********************************************************************
 *net_lib.c : Implements the function definitions of net_lib.h, which
 *            are for use with netmon.c
 *
 *Author:    Dwayne Hoy / DoogeJ
 *
 *Date:      September 09 2005
 *
 *Version:   0.5.2
 *
 *Modified:  December 30 2015
 **********************************************************************/

#include "net_lib.h"


void display(net_t *Ndata, int size ,double time){
  int i;
  long double tx;          //Stores the tx total
  long double rx;          //Stores the rx total
  long double rx_trans;    //Stores the rx rate
  long double tx_trans;    //Stores the tx rate
  char tx_unit[4];         //Unit to display tx in
  char rx_unit[4];         //Unit to display rx in
  char rx_trans_unit[4];   //Unit to display rx rate in
  char tx_trans_unit[4];   //Unit to display tx rate in
  if(!DEBUG)
    clearscreen();
  /***********Prints title header*********/
  fprintf(stdout, "%s %16s %14s %11s %14s\n", "Device", "Downloaded",
	  "Rate" ,"Uploaded", "Rate");

  for(i = 0; i < size; i++){
    tx = to_unit(tx_unit, get_txbtotal(&Ndata[i]),
		 get_txgtotal(&Ndata[i]), 1);
    rx = to_unit(rx_unit, get_rxbtotal(&Ndata[i]),
		 get_rxgtotal(&Ndata[i]), 1);

    tx_trans = get_txRate(&Ndata[i], time);
    tx_trans = to_unit(tx_trans_unit, tx_trans, 0, 0);
    rx_trans = get_rxRate(&Ndata[i], time);
    rx_trans = to_unit(rx_trans_unit, rx_trans, 0, 0);

    fprintf(stdout, "%4s %15.2Lf %2s %9.2Lf %2s/s %8.2Lf %2s %9.2Lf %2s/s\n",
	    Ndata[i].device , rx, rx_unit, rx_trans, rx_trans_unit, tx,
	    tx_unit, tx_trans, tx_trans_unit);

  }
}

void clearscreen(){
  char clear[3];
  clear[0] = 033; clear[1] = 'c'; clear[2]='\0';
  fprintf(stdout, "%s",clear);
}



long double to_unit(char *unit, long double bval, long double gval,
		    int mode){
  long double tval;
  if(bval < 1024 && gval == 0){
    strcpy(unit, "B");
    return bval;
  }
  else if(bval < 1048576 && gval == 0){
    strcpy(unit, "KB");
    tval = (bval / 1024);
    return tval;
  }
  else if(bval < 1073741824 && gval == 0){
    strcpy(unit, "MB");
    tval = (bval / 1048576);
    return tval;
  }
  else{
    strcpy(unit, "GB");
    tval = gval;
    tval += (bval / (long double)1073741824);
    return (tval);
  }
}



int checkfile(char *path){
  struct stat attrib;
  if((lstat(path, &attrib)) != 0){
    if(DEBUG)
      fprintf(stderr, "checkfile: Error: Unable to get file statistics\n");
    return -1;
  }
  if(S_ISLNK(attrib.st_mode)){
    if(DEBUG)
      fprintf(stderr, "checkfile: Error: File is a symbolic link\n");
    return -2;
  }
  if(DEBUG)
    fprintf(stderr, "checkfile: file security check has passed\n");
  return 0;
}

int ver_check(FILE *dev_file){
  char buff[80];
  int i = 0;

  for(i = 0; i < 2; i++){
    if(!fgets(buff, sizeof(buff), dev_file)){
      if(DEBUG)
        fprintf(stderr, "ver_check: Error: Unable to read dev file\n");
      return -1;
    }
  }

  if((strstr(buff ,"compressed")) == NULL){
    if(DEBUG)
      fprintf(stderr, "ver_check: Error: Incorrect dev version\n");
    return -2;
  }
  if(DEBUG)
    fprintf(stderr, "ver_check: dev version is correct\n");
  return 0;
}

void remove_newline(char* string){
  int i = 0;
  for(i = 0; i < strlen(string); i++){
    if(string[i] == '\n'){
      string[i] = '\0';
      if(DEBUG)
	fprintf(stderr, "remove_newline: Successfully removed a new line\n");
      return;
    }
  }
}

int write_to_log(net_t *Ndata, char* path, int size){
  FILE *log;
  char temp_string[300];
  int i;
  if(((log = fopen(path, "w")) == NULL))
    return -1;
  if(DEBUG)
    fprintf(stderr, "write_to_log: Successfully opened file for writting\n");
  if(!fputs("<0.5>\n", log)){
    fprintf(stderr, "write_to_log: Error writting to log\n");
    return -1;
  }

  for(i = 0; i < size; i++){
    if((snprintf(temp_string, 300, "[%s]\n%Lu\n%Lu\n%Lu\n%Lu\n",
		 Ndata[i].device,
		 get_rxbtotal(&Ndata[i]),
		 get_rxgtotal(&Ndata[i]),
		 get_txbtotal(&Ndata[i]),
		 get_txgtotal(&Ndata[i]))) < 0 ){
      fprintf(stderr, "write_to_log: Error creating string");
      return -1;
    }

    if(DEBUG)
      fprintf(stderr, "write_to_log: Successfully created string\n");

    if(!fputs(temp_string, log)){
      fprintf(stderr, "write_to_log: Error writting to log\n");
      return -1;
    }

    if(DEBUG)
      fprintf(stderr, "write_to_log: Successfully wrote to log\n");
  }
  fclose(log);
  return 0;
}


int parse_args(int argc, char *argv[]){
  int i;
  if(argc == 1)
    return 0;
  for(i = 1; i < argc; i++){
    if(strcmp(argv[i], "--config") == 0)
      return 1;
    else if(strcmp(argv[i], "--help") == 0){
      display_help();
      return -1;
    }
    else{
      fprintf(stderr, "Invalid argument please use --help flag for more details\n");
      return -1;
    }

  }
  return 0;
}

void display_help(){
  fprintf(stdout, "Usage netmonitor [OPTION]\n");
  fprintf(stdout, "A network monitoring console application.\n\n");
  fprintf(stdout, "  --help         Displays this text\n");
  fprintf(stdout, "  --config       Runs the configuration wizard\n\n");
  fprintf(stdout, "Project page: https://github.com/DoogeJ/netmonitor-ng\n\n");
}

